src/libraries/System.Private.CoreLib/src/System/Diagnostics/Tracing/EventSource.cs

```diff
@@ -1692,127 +1692,150 @@ private static Guid GenerateGuidFromName(string name)
             return new Guid(bytes);
         }
 
-        private static unsafe object? DecodeObject(Type dataType, ref EventSource.EventData* data)
+        private static unsafe object? DecodeObject(Type dataType, ref EventData* data)
         {
             // TODO FIX : We use reflection which in turn uses EventSource, right now we carefully avoid
             // the recursion, but can we do this in a robust way?
 
             IntPtr dataPointer = data->DataPointer;
-            // advance to next EventData in array
-            ++data;
+            Debug.Assert(dataPointer != IntPtr.Zero || data->Size == 0);
 
-            Again:
-            if (dataType == typeof(IntPtr))
+            if (dataType == typeof(string))
             {
-                return *((IntPtr*)dataPointer);
+                goto String;
             }
             else if (dataType == typeof(int))
             {
-                return *((int*)dataPointer);
+                Debug.Assert(data->Size == 4);
+                return *(int*)dataPointer;
             }
-            else if (dataType == typeof(uint))
-            {
-                return *((uint*)dataPointer);
-            }
-            else if (dataType == typeof(long))
-            {
-                return *((long*)dataPointer);
-            }
-            else if (dataType == typeof(ulong))
-            {
-                return *((ulong*)dataPointer);
-            }
-            else if (dataType == typeof(byte))
-            {
-                return *((byte*)dataPointer);
-            }
-            else if (dataType == typeof(sbyte))
-            {
-                return *((sbyte*)dataPointer);
-            }
-            else if (dataType == typeof(short))
-            {
-                return *((short*)dataPointer);
-            }
-            else if (dataType == typeof(ushort))
-            {
-                return *((ushort*)dataPointer);
-            }
-            else if (dataType == typeof(float))
-            {
-                return *((float*)dataPointer);
-            }
-            else if (dataType == typeof(double))
-            {
-                return *((double*)dataPointer);
-            }
-            else if (dataType == typeof(decimal))
-            {
-                return *((decimal*)dataPointer);
-            }
-            else if (dataType == typeof(bool))
-            {
-                // The manifest defines a bool as a 32bit type (WIN32 BOOL), not 1 bit as CLR Does.
-                return *((int*)dataPointer) == 1;
-            }
-            else if (dataType == typeof(Guid))
-            {
-                return *((Guid*)dataPointer);
-            }
-            else if (dataType == typeof(char))
-            {
-                return *((char*)dataPointer);
-            }
-            else if (dataType == typeof(DateTime))
+
+            TypeCode typeCode = Type.GetTypeCode(dataType);
+            int size = data->Size;
+
+            if (size == 4)
```

Comment:
Wouldn't be a simple switch over typeCode instead of multiple levels of nested ifs faster?

Reply:
They don't always match up here.
For enums, EventSource treats them as `int32`, but `Type.GetTypeCode(dataType)` will resolve the underlying type.
That is why the `[SByte, Int32]` range is being checked under `size == 4` (and read as an `int*`), but then also checked outside the `size == 4` case.

Reply:
Unless you'd like to optimize for non-int enums it still seems to me that the switch would be better option

Reply:
In that case you would have duplication like
```c#
case TypeCode.Byte:
        if (dataType.IsEnum)
        {
            Debug.Assert(size == 4);
            return *(int*)dataPointer;
        }
        else
        {
            Debug.Assert(size == 1);
            return *(byte*)dataPointer;
        }

case TypeCode.SByte:
        if (dataType.IsEnum)
        {
            Debug.Assert(size == 4);
            return *(int*)dataPointer;
        }
        else
        {
            Debug.Assert(size == 1);
            return *(sbyte*)dataPointer;
        }
// ...
```
For byte, sbyte, short, ushort, int.

I can measure, but I doubt changing to a switch would be faster here.